### ISSUE TITLE
RUF029 triggers on FastAPI routes when they're inner functions

### ISSUE THREAD

**Issue by scy** [astral-sh/ruff#14903]
The following code triggers [`RUF029` (unused async)](https://docs.astral.sh/ruff/rules/unused-async/):

```python
from fastapi import FastAPI


def setup_app(app: FastAPI) -> None:
    @app.get("/")
    async def get_root() -> str:  # RUF029 here
        return "Hello World!"


app = FastAPI()


@app.get("/foo")
async def get_foo() -> str:  # no error
    return "Hello this is foo"
```

Note that it doesn't trigger on the top-level function.

That's because there is special handling in the code for FastAPI, implemented by @TomerBin in #12925/#12938, but I guess the underlying cause for the different outcome is that one of the `is_fastapi_*` checks in `rules/fastapi/rules/mod.rs` doesn't detect this case correctly.

**Comment by TomerBin** [comment]
Hi @scy 
Thanks for pointing that out :)
A complete solution will depend on a proper type checking in ruff, but until it'll be landed I can improve the current logic to also detect fastapi routes whose app comes from a function typed argument.

Hopefully get into that in the following days :)
